pkg/mimirtool/commands/remote_read.go

```diff
@@ -449,3 +489,108 @@ func (c *RemoteReadCommand) export(_ *kingpin.ParseContext) error {
 
 	return nil
 }
+
+func (c *RemoteReadCommand) remoteWrite(_ *kingpin.ParseContext) error {
+	write, err := c.writeClient()
+	if err != nil {
+		return fmt.Errorf("creating remote write client: %w", err)
+	}
+
+	query, _, _, err := c.prepare()
+	if err != nil {
+		return err
+	}
+
+	timeseries, err := query(context.Background())
+	if err != nil {
+		return err
+	}
+
+	samples, exemplars, histograms := 0, 0, 0
+	for _, ts := range timeseries {
+		samples += len(ts.Samples)
+		exemplars += len(ts.Exemplars)
+		histograms += len(ts.Histograms)
+	}
+	log.Infof("Got %d samples, %d exemplars, %d histograms for %d timeseries", samples, exemplars, histograms, len(timeseries))
+
+	req := prompb.WriteRequest{Timeseries: make([]prompb.TimeSeries, 0, c.write.batchSize)}
+	compressed := make([]byte, 0, 1024*1024)
+	pbuf := proto.NewBuffer(nil)
+	for idx, ts := range timeseries {
+		req.Timeseries = append(req.Timeseries, *ts)
+
+		if idx == 0 || idx%c.write.batchSize != 0 {
+			continue
+		}
+
+		// Send a batch.
+		pbuf.Reset()
+		if err := pbuf.Marshal(&req); err != nil {
+			return err
+		}
+
+		compressed = snappy.Encode(compressed[:0], pbuf.Bytes())
```

Comment:
IIUC, this will panic if `pbuf.Bytes()` is > 4GB. You may want to check for a negative return value from `snappy.MaxEncodedLen` before attempting to encode.

Reply:
While true, I'd say it improbable that the remote read instance would return so many samples in the first place. :D